beacon_node/eth2_libp2p/src/discovery/mod.rs

```diff
@@ -422,26 +426,31 @@ impl<TSpec: EthSpec> Discovery<TSpec> {
 
         // Check that we are within our query concurrency limit
         while !self.at_capacity() && !self.queued_queries.is_empty() {
+            // use this to group subnet queries together for a single discovery request
+            let mut subnet_queries: Vec<SubnetQuery> = Vec::new();
+
             // consume and process the query queue
             match self.queued_queries.pop_front() {
                 Some(QueryType::FindPeers) => {
-                    // Only permit one FindPeers query at a time
-                    if self.find_peer_active {
-                        self.queued_queries.push_back(QueryType::FindPeers);
-                        continue;
-                    }
                     // This is a regular request to find additional peers
                     debug!(self.log, "Discovery query started");
                     self.find_peer_active = true;
-                    self.start_query(QueryType::FindPeers, FIND_NODE_QUERY_CLOSEST_PEERS);
+                    self.start_query(
+                        vec![QueryType::FindPeers],
```

Comment:
If we change the structure of `QueryType` this would not need to be a vec. 

I think we should still prioritise whatever query is first. Queries take a long time and subnet queries are important for a validator to make money, whereas FindPeers queries and passive queries just to get to a specific peer count. To save complexity, we could just `pop_back()` the find_node queries when going through the queue. 

Something like 
[Sub1, FP, Sub2, Sub3, Sub4] 
with Subx = subnet query for subnet x and FP = findpeers. 
We should go through, grab Sub1, put the FP on the back, grab Sub and Sub3 and process [Sub1, Sub2, Sub3]. 
The loop should then see [Sub4, FP]. 
It should grab sub4, check for any more subs, if none, then process sub4 leaving the queue [FP]. If there are enough slots, then process FP. 

Hopefully this logic isn't too complicated. I think it's important enough to prioritise Sub queries over FP queries to warrant the complexity.

Reply:
> If we change the structure of QueryType this would not need to be a vec.

I created a new `GroupedQueryType` with a slightly different structure that helped clean up a lot of stuff. 

> I think we should still prioritise whatever query is first. Queries take a long time and subnet queries are important for a validator to make money, whereas FindPeers queries and passive queries just to get to a specific peer count...

I updated it to what I think you described. But I was a little unclear about how this scenario should work:

[Sub1, Sub2, Sub3, FP, Sub4]

If we want to query in groups of three do we want to:
1. Request<1,2,3>,  Request<FP>, Request<4>

Or:
2. Request<1,2,3>,  Request<4>, Request<FP>

Reply:
hmm good point. I was originally thinking 1, but now that you raise it, 2 would be better. 

Yeah, it might make is simpler if we just prioritise all subnet queries